release-20.1: cli/zip: misc fixes

knz opened the pull request:
Backport 1/1 commits from #48073.
Backport 1/1 commits from #48096.
Backport 1/1 commits from #48099.
Backport 1/1 commits from #48109.
Backport 1/1 commits from #48150. 

/cc @cockroachdb/release

---

knz committed:
cli/zip: skip over decommissioned nodes

Release note (cli change): `debug zip` now properly skips over fully
decommissioned nodes.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/48094)
<!-- Reviewable:end -->

tbg reviewed (approved):
I know you're aware, but hold off on the merge.

knz committed:
cli/zip: avoid buffering SQL data in RAM

Previously, `cockroach debug zip` would fetch all SQL rows in RAM:

- a first time into an array of strings, to determine the maximum
  table width in the "pretty" ASCII-art formatter,
- then a second time in a `bytes.Buffer`,

before being eventually streamed out into the output ZIP file.

This was causing problems when extracting very large `system` tables,
for example out of runaway job, event log or range log tables.

This patch fixes it by switching to a streaming SQL row renderer (TSV)
and avoiding the intermediate byte buffer.

A side benefit is that the TSV format is arguably much more easy to
scan to the human eye when the tables are very wide.

Release note (backward-incompatible change): The copy of system and
`crdb_internal` tables extracted by `cockroach debug zip` is now
written using the TSV format (inside the ZIP file), instead of an
ASCII-art table as previously.

Release note (bug fix): `cockroach debug zip` can now successfully
avoid out-of-memory errors when extracting very large `system` or
`crdb_internal` tables.

knz committed:
cli/zip: loop upon SQL retries

If a SQL query in `cockroach debug zip` matches the following
conditions:

- returns more than N rows, where N is the configured server-side row
  buffer for server-side retries,
- encounters a retry error,

then previous to this patch it would just bail and fail entirely.

This patch adds a modest retry loop in this case (5 iterations).

Release note (cli change): `cockroach debug zip` now tries multiple
times to retrieve data using SQL if it encounters retry errors.

knz committed:
cli/zip: do not ignore close errors

Release note (bug fix): `cockroach debug zip` will now properly report
an error if some error is encountered while writing the end of the
output ZIP file.

knz committed:
cli/zip: makes it possible to include/exclude specific nodes

Release note (cli change): `cockroach debug zip` now supports two
command-line parameters `--nodes` and `--exclude-nodes`. When
specified, they control which nodes are inspected when gathering the
data. This makes it possible to focus on a group of nodes of interest
in a large cluster, or to exclude nodes that `debug zip` would have
trouble reaching otherwise. Both flags accept a list of individual
node IDs or ranges of node IDs, e.g. `--nodes=1,10,13-15`.
